doubly_linked_list/doubly_linked_list.py

```diff
@@ -44,25 +44,102 @@ def __len__(self):
     return self.length
 
   def add_to_head(self, value):
-    pass
-
+    self.length +=1
+    new_node = ListNode(value = value)
+    new_node.next = None
+    new_node.prev = None
+    if self.head is None:
+      self.head = new_node
+      self.tail = new_node
+    else:
+      self.head.prev = new_node
+      new_node.next = self.head
+      self.head = new_node
+    
   def remove_from_head(self):
-    pass
+    if self.head == None:
+      return
+    else:
+      self.length -=1
+      if self.head != self.tail:
+        self.head = self.head.next
+        self.head.prev = None
+      else:
+        self.head = self.tail=None
 
   def add_to_tail(self, value):
-    pass
+    self.length +=1
+    new_node = ListNode(value=value)
+    new_node.next = None
+    new_node.prev = None
+    if self.tail is None:
+      self.head = new_node
+      self.tail = new_node
+    else:
+      self.tail.next = new_node
+      new_node.prev = self.tail
+      self.tail = new_node
 
+   
   def remove_from_tail(self):
-    pass
+    if self.head is None:
+      return
+    if self.head.next is None:
+      self.head = None
+      return
+    n = self.head
+    while n.next is not None:
+      n = n.next
+    n.prev.next = None
+
+    
 
   def move_to_front(self, node):
-    pass
+    # get to the front node (there won't be a previous node)
+    curr_node = self
+    while curr_node.prev is not None:
+      print("moving to the prev node")
+      curr_node = curr_node.prev
+    # then add the node to its head
+    curr_node.add_to_head(node)
+   
```

Comment:
this was also giving you problems. I wrote this:

if node is self.head:
____return
value = node.value
if node is self.tail:
____self.remove_from_tail()
else:
____node.delete()
____self.length -= 1
self.add_to_head(value)

you have helper functions above so you can utilize those if you need!